In [ ]:
!cat /proc/cpuinfo

processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 79
model name	: Intel(R) Xeon(R) CPU @ 2.20GHz
stepping	: 0
microcode	: 0xffffffff
cpu MHz		: 2199.998
cache size	: 56320 KB
physical id	: 0
siblings	: 2
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs taa mmio_stale_data retbleed
bogomips	: 4399.99
clflush size	: 64
cache_alignment	: 64
addres

# I Solving the Sum of Two Digits Programming Challenge

## C++

In [ ]:
%%writefile APlusB.cpp
#include <iostream>

int sum_of_two_digits(int first_digit, int second_digit) {
    return first_digit + second_digit;
}

int main() {
    int a = 0;
    int b = 0;
    std::cin >> a;
    std::cin >> b;
    std::cout << sum_of_two_digits(a, b);
    return 0;
}

Writing APlusB.cpp


In [17]:
!g++ APlusB.cpp -o APlusB -std=c++11
!./APlusB

^C


### Test input output files

In [ ]:
%%writefile input1.txt
4 8

Writing input1.txt


In [ ]:
%%writefile output1.txt
12

Writing output1.txt


In [ ]:
%%writefile input2.txt
0 9

Writing input2.txt


In [ ]:
%%writefile output2.txt
9

Writing output2.txt


In [ ]:
!./APlusB < input1.txt

12

In [ ]:
!./APlusB < input2.txt

9

### Comparing files with grep

https://askubuntu.com/questions/546796/comparing-two-text-files

In [ ]:
!./APlusB < input1.txt> ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [ ]:
!./APlusB < input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

## Python

In [ ]:
%%writefile sum.py
# python3

def sum_of_two_digits(first_digit, second_digit):
    return first_digit + second_digit

if __name__ == '__main__':
    a, b = map(int, input().split())
    print(sum_of_two_digits(a, b))

Writing sum.py


In [ ]:
!python sum.py

Traceback (most recent call last):
  File "/content/sum.py", line 7, in <module>
    a, b = map(int, input().split())
KeyboardInterrupt
^C


In [ ]:
!python sum.py  < input1.txt> ptest1.txt
!grep -xvFf output1.txt ptest1.txt

In [ ]:
!python sum.py  < input2.txt> ptest2.txt
!grep -xvFf output2.txt ptest2.txt

# II Solving The Maximum Pairwise Product Programming Challenge in C++

## C++

$$ T(n) < 6 n^2 = O(n^2)$$

In [ ]:
%%writefile max_pairwise_product_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int aux_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second ){
                aux_product = numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product_full.cpp


#### Time Complexity
$$\left. \begin{array}{llc}
 \text{INSTRUCTION} & \text{TYPE}& \text{# EXECUTIONS}\\
\text{first != second} &  \text{comparisons} &n^2 \\
\text{numbers[first] * numbers[second]}& \ \text{ products}  &n^2 - n \\
\text{max_product = aux_product} &  \text{assigments} & [1, \frac{n^2 - n}{2})  
\end{array}\right\}  =  O(n^2)$$

In [ ]:
!g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
!./max_pairwise_product_full

^C


### Test 1: Small numbers

In [18]:
%%writefile input0.txt
10
0 0 0 0 0 0 0 0 0 0

Writing input0.txt


In [19]:
%%writefile output0.txt
0

Writing output0.txt


In [20]:
!./max_pairwise_product_full< input0.txt > ctest0.txt
!grep -xvFf output0.txt ctest0.txt

In [21]:
%%writefile input1.txt
10
1 2 3 4 5 6 7 8 9 10

Overwriting input1.txt


In [22]:
!./max_pairwise_product < input1.txt

/bin/bash: line 1: ./max_pairwise_product: No such file or directory


In [23]:
%%writefile output1.txt
90

Overwriting output1.txt


In [24]:
!./max_pairwise_product_full< input1.txt > ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [25]:
%%writefile input2.txt
10
10 9 8 7 6 5 4 3 2 1

Overwriting input2.txt


In [26]:
%%writefile output2.txt
90

Overwriting output2.txt


In [27]:
!./max_pairwise_product_full< input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

### Test 2: largest input numbers 200000 200000 to check possible intger overflow

In [28]:
%%writefile input3.txt
6
1 3 200000 9 200000 23

Writing input3.txt


In [29]:
200000*200000

40000000000

In [30]:
%%writefile output3.txt
40000000000

Writing output3.txt


In [31]:
!./max_pairwise_product_full < input3.txt

1345294336


In [32]:
!./max_pairwise_product_full < input3.txt > ctest3.txt
!grep -xvFf output3.txt ctest3.txt

1345294336


In [33]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
for i in {0..3}
do
    ./max_pairwise_product_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Writing run_all_test.sh


In [34]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

1345294336


### Changing the variables to long long

In [35]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second){
                aux_product = (long long) numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }


    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Writing max_pairwise_product_long_full.cpp


In [36]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full

In [37]:
!./max_pairwise_product_long_full < input3.txt

40000000000


### Runing multuple test with a bash shell

In [38]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
for i in {1..3}
do
    ./max_pairwise_product_long_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


In [39]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

### Test 3  stress tests, 6 testing cases files input4.txt/output4.txt .. input10.txt/output10.txt

---



In [40]:
# Stress Test case 1
%%writefile input4.txt
6
1000000000 234000000 200000000 9000 200000000 892000000

Writing input4.txt


In [41]:
1000000000*892000000


892000000000000000

In [42]:
%%writefile output4.txt
892000000000000000

Writing output4.txt


In [43]:
#Stress Test case 2
%%writefile input5.txt
# 5 random negative numbers
5
-1000000000 -234000000 -200000000 -9000 -200000000

Writing input5.txt


In [44]:
-1000000000*-234000000

234000000000000000

In [45]:
%%writefile output5.txt
234000000000000000

Writing output5.txt


In [46]:
#Stress Test case 3
%%writefile input6.txt
# 5 random numbers
5
10000000 4 200000 9000 1

Writing input6.txt


In [47]:
10000000*200000

2000000000000

In [48]:
%%writefile output6.txt
2000000000000

Writing output6.txt


### Test 4  time test  measure the execution time running with the the largest worst cases input11.txt  (200000 increasing numbers) and input12.txt (199998 increasing numbers and  200000, 200000 at the end)

In [49]:
# input11.txt 200000 increasing numbers:  1 2 .. 200000

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input11.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Writing WriteTest.cpp


In [50]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [51]:
199999 * 200000

39999800000

In [53]:
%%writefile output11.txt
39999800000

Writing output11.txt


In [52]:
# input12.txt 200000 decreasing numbers and 200000, 199999 ...  2 1

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input12.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << 200001-i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [54]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [55]:
%%writefile output12.txt
39999800000

Writing output12.txt


### Measure time

https://askubuntu.com/questions/53444/how-can-i-measure-the-execution-time-of-a-terminal-process

In [57]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input11.txt

^C


In [58]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input12.txt

^C


### Calculate the number or and upper bound to the number of instructions executed by the algorithm as a function of n, add in each line a comment with the number of instructions that are executed in that line.

In [59]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //INSTRUCTIONS executed by line
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //2+2n
        for (int second = 0; second < n; ++second) {                         //2n + 2n**2
          if ( first != second ){                                            //n**2
                aux_product = numbers[first] * numbers[second];              //n**2 -n
                if (aux_product > max_product)                               //n**2 -n
                    max_product = aux_product;                               //[0, n**2/2 - n/2] upper bound
          }
        }
    }

                                                                             // --------------
    return max_product;                                                      // 5 + 2*n + [5*n**2, 11*n**2/2  - n/2]  upper bound
}

int main() {                                              //INSTRUCTIONS executed by line
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //2 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 5 + 2*n + [5*n**2, 11*n**2/2  - n/2]

    return 0;                                             // --------------
}                                                         //  11 + 6*n + [5*n**2, 11*n**2/2  - n/2]  upper bound


Overwriting max_pairwise_product_long_full.cpp


### Count executed intructions in code

In [60]:
%%writefile max_pairwise_product_long_full_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {      //INSTRUCTIONS executed by line

    long long max_product = 0LL;                                                          //1
    execints++;

	  long long aux_product = 0LL;                                                          //1
    execints++;

    int n = numbers.size();                                                               //1
    execints++;

    execints++;
    for (int first = 0; first < n; ++first) {                                             //2+2n
        execints += 2;

        execints++;
        for (int second = 0; second < n; ++second) {                                       //2n + 2n**2
          execints += 2;

          execints++;
          if ( first != second ){                                                         //n**2
                aux_product = numbers[first] * numbers[second];                           //n**2 -n
                execints++;

                execints++;
                if (aux_product > max_product){                                           //n**2 -n
                    max_product = aux_product;                                            //[0,n**2/2 - n/2)  upper bound
                    execints++;

                }
        }

      }
      execints++;
    }
    execints++;
                                                                                        // --------------
    return max_product;                                                                 // 5 + 2*n + [5*n**2, 11*n**2/2  - n/2] upper bound
}

int main() {                                                 //INSTRUCTIONS executed by line

    long long execints =0LL;

    int n;                                                   //1
    execints++;

    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                             //n
    execints += n;

    execints++;
    for (int i = 0; i < n; ++i) {                            //2 + 2n
        execints += 2;

        std::cin >> numbers[i];
        execints++;                                          //n

    }


    //std::cout << MaxPairwiseProduct(numbers) << "\n";      //1 + 1  + 5 + 2*n + [5*n**2, 11*n**2/2  - n/2] upper bound

    execints++;
    int M = MaxPairwiseProduct(numbers,execints);

    execints++;
    std::cout <<  execints << "\n";
    return 0;                                                // --------------
}                                                            // 11 + 6*n + [5*n**2, 11*n**2/2  - n/2]

Writing max_pairwise_product_long_full_execints.cpp


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T (n) = 11 + 6n + \left[5n^2, 11\frac{n^2}{2}  - \frac{n}{2}\right)  >  6n^2 = O(n^2)$$

It depends on the size of the lists and  the order of the list.

For the input0.txt
```
10
0 0 0 0 0 0 0 0 0 0
```
$$T = 11 + 6(10) + 5(10)^2   = 561 $$

$$T(10) \in [571, 621)  $$


In [61]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input0.txt

570


For the input1.txt
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 11 + 6(10) + 5(10)^2 +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0  = 601 $$

$$T(10) \in [572, 622)  $$


In [ ]:
#The number of times that  max_product = aux_product; in the worst case is less than (n**2 -n)/2 -  upper bound

#     1  2  3  4  5  6  7  8  9  10
#  1  x  2  3  4  5  6  7  8  9  10         9
#  2  2  x  6  8  10 12 14 16 18 20         5 14
#  3  3  6  x  12 15 18 21 24 27 30         4 18
#  4  4  8  12 x  20 24 28 32 36 40         3 21
#  5  5  10 15 20 x  30 35 40 45 50         2 23
#  6  6  12 18 24 30 x  42 48 54 60         2 25
#  7  7  14 21 28 35 42 x  56 63 70         2 27
#  8  8  18 27 36 45 54 63 72 x  90         1 30
#  10 10 20 30 40 50 60 70 60 90 x          0

In [62]:
11 + 6*(10) + 5*(10)*(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0

601

In [63]:
n = 10
print( "[", 11 + 6*n + 5*n**2,  ",", 11 + 6*n  + 5.5*n**2, ")")

[ 571 , 621.0 )


In [64]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input1.txt

600



For the input2.txt
```
10
10 9 8 7 6 5 4 3 2 1
```
$$T = 11 + 6*(10) + 5*(10)**2 +  1  = 571 $$

In [65]:
11 + 6*(10) + 5*(10)**2 +  1

572

In [66]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input2.txt

571


Calculate estimated time by instruction and number of instructions by second

For the input11.txt
```
200000
1 2 ..  199999 2000
```
$$T = 11 + 6*(200000) + 5.5*(200000)**2  = 200001200012  $$

In [67]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input11.txt

^C


In [68]:
# Estimated time by instruction
# use the real reported  !time ./max_pairwise_product_long_full< input11.txt in seconds intead of the 295

295/200003177092


1.4749765693187072e-09

In [69]:
# Estimated number of instructions by second

200003177092/295

677976871.4983051

### Estimate and verify the maximun n that can be solved in 1 second

$$(5.5n^2 + 5.5n + 11) \times 1.47*10^{-9} s = 1 s$$

$$ n \approx 11000$$


In [70]:
# Verify input13.txt
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input13.txt");

 MyFile << 11000 <<"\n";  // n
 for (int i=1; i <11001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [71]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [72]:
10999 * 11000

120989000

In [73]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input13.txt

120989000

real	0m0.865s
user	0m0.845s
sys	0m0.004s


## Python

In [74]:
%%writefile max_pairwise_product_long_full.py
def MaxPairwiseProduct(numbers):
  max_product = 0
  aux_product = 0
  n = len(numbers)

  for first in range (n):
    for second in range(n):
      if(first != second):
        aux_product = numbers[first]*numbers[second]
        if aux_product > max_product:
          max_product = aux_product
  return max_product

n = int(input())
numbers = [int(i) for i in input().split()]
print(MaxPairwiseProduct(numbers))

Writing max_pairwise_product_long_full.py


In [75]:
# Verify
!time python3 max_pairwise_product_long_full.py < input13.txt

Traceback (most recent call last):
  File "/content/max_pairwise_product_long_full.py", line 16, in <module>
    print(MaxPairwiseProduct(numbers))
  File "/content/max_pairwise_product_long_full.py", line 10, in MaxPairwiseProduct
    if aux_product > max_product:
KeyboardInterrupt
^C


# For the follwing solutions rquired  the same tests and analysis performed for the first solution

## Solutions in C++ and Python that uses
$$\frac{n^2 - n}{2} = O(n^2)\ \ \text{products}$$

In [76]:
%%writefile max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product.cpp


In [77]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //INSTRUCTIONS executed by line
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //2+2n
        for (int second = first + 1; second < n; ++second) {                 //2n + (n-1)n
			       aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2
            if (max_product < aux_product)  {                                //(n-1)n/2
				        max_product = aux_product;                                   //[1,(n-1)n/2)  upper bound
             }
        }
    }
                                                                             // --------------
    return max_product;                                                      // 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
}

int main() {                                              //INSTRUCTIONS executed by line
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //2 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
    return 0;                                             // --------------
}                                                         // 11 + 7n + [2(n-1)n+1, 2.5 (n-1)n) upper bound


Writing max_pairwise_product_long.cpp


In [78]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!time ./max_pairwise_product_long < input13.txt

120989000

real	0m0.397s
user	0m0.389s
sys	0m0.004s


### Estimate and verify the maximun n that can be solved in 1 second



# Sort solution in C++ and Python

$$ O(n \lg n)$$

**Which sorting algorithm is used in GCC?**
https://stackoverflow.com/questions/7221257/which-sorting-algorithm-is-used-in-gcc

* ...  the algorithm is called introsort, a combination between Insertionsort, Quicksort & Heapsort. ...

**Introsort**
https://en.wikipedia.org/wiki/Introsort

Introsort or introspective sort is a hybrid sorting algorithm that provides both fast average performance and (asymptotically) optimal worst-case performance. It begins with quicksort, it switches to heapsort when the recursion depth exceeds a level based on (the logarithm of) the number of elements being sorted and it switches to insertion sort when the number of elements is below some threshold

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Average performance	$O(n \lg n)$

**Timsort**
https://en.wikipedia.org/wiki/Timsort

Timsort is a hybrid, stable sorting algorithm, derived from merge sort and insertion sort, designed to perform well on many kinds of real-world data. It was implemented by Tim Peters in 2002 for use in the Python programming language.

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Best-case performance $O(n)$
*   Average performance	$O(n \lg n)$
*   Worst-case space complexity $O(n)$

In [79]:
# %%writefile max_pairwise_product_with_sort.py
# #Python3
# if __name__ == "__main__":
#     n = int(input())
#     lista = list(map(int, input().split()))
#     lista.sort()
#     print(lista[n - 1] * lista[n - 2])


%%writefile max_pairwise_product_sort.py
def max_pairwise_product_sort(numbers):
  numbers.sort(reverse=True)
  return numbers[0] * numbers[1]

n = int(input())
numbers = [int(i) for i in input().split()]
print(max_pairwise_product_sort(numbers))

Writing max_pairwise_product_sort.py


Test with inputs 1 and 2



In [80]:
# !./max_pairwise_product_with_sort_C < input1.txt
# !./max_pairwise_product_with_sort_C < input2.txt


%%writefile run_all_test.sh
#comparing all the executions to check the differences between the outputs
for i in {1..2}
do
    python3 max_pairwise_product_sort.py < input$i.txt > pythonOutput$i.txt
    grep -xvFf pythonOutput$i.txt output$i.txt
done

Overwriting run_all_test.sh


Test and time it with input 3

In [81]:
!time python3 max_pairwise_product_sort.py < input3.txt > pythonOutput3.txt
!grep -xvFf pythonOutput3.txt output3.txt


real	0m0.035s
user	0m0.024s
sys	0m0.010s


20 stress tests

In [82]:
import random

def max_pairwise_efficient(numbers):
    n = len(numbers)
    biggest_1 = float('-inf')
    biggest_2 = float('-inf')

    for i in range(n):
        new = numbers[i]
        temp = biggest_1
        if new > biggest_1:
          biggest_2 = biggest_1
          biggest_1 = new
        if temp > biggest_2: biggest_2 = temp

    return biggest_1 * biggest_2

# 20 stress tests
for _ in range(20):
    n = random.randint(2, 12)  # Use a large n
    numbers = [random.randint(1, 10**9) for _ in range(n)]  # Generate n random numbers between 1 and 10^9

    print("-------------------------------------------------")
    print(f"Testing with the list: {numbers}")
    print(f"The efficient solution gives the result {max_pairwise_efficient(numbers)}")
    # Add any additional test or comparison here if needed
    print("-------------------------------------------------")

-------------------------------------------------
Testing with the list: [990257951, 659184331, 975047833]
The efficient solution gives the result 980610809518718401
-------------------------------------------------
-------------------------------------------------
Testing with the list: [216552260, 536373975, 74646221, 606749043, 884542484, 83697596, 863815999, 368242898]
The efficient solution gives the result 782415406000890256
-------------------------------------------------
-------------------------------------------------
Testing with the list: [574185287, 312292060, 289042034, 934755079]
The efficient solution gives the result 536722613310322673
-------------------------------------------------
-------------------------------------------------
Testing with the list: [143324073, 902683616, 677490945, 478241968, 221426798, 632327997, 758206227, 268739361]
The efficient solution gives the result 814837710594835456
-------------------------------------------------
-----------------

Count executed instructions in code and verify with inputs 1,2 and 3


In [83]:

%%writefile max_pairwise_efficient.py
import random
execints = 0

def max_pairwise_efficient(numbers):
    global execints
    n = len(numbers)  # Tamaño de la lista
    biggest_1 = float('-inf')  # Inicializa el valor más grande en -infinito
    biggest_2 = float('-inf')  # Inicializa el segundo valor más grande en -infinito
    execints += 3  # Se cuenta como 3 instrucciones iniciales

    for i in range(n):
        new_number = numbers[i]  # Toma el siguiente número de la lista
        temp = biggest_1  # Almacena temporalmente el valor más grande encontrado
        execints += 2  # 2 instrucciones por la asignación y la comparación

        if new_number > biggest_1:  # Si el nuevo número es mayor que el más grande
            biggest_2 = biggest_1  # El anterior mayor pasa a ser el segundo más grande
            biggest_1 = new_number  # El nuevo número se convierte en el más grande
            execints += 2  # 2 instrucciones por el bloque if

        if temp > biggest_2:  # Si el valor temporal (anterior mayor) es mayor que el segundo mayor
            biggest_2 = temp  # Se actualiza el segundo mayor
            execints += 2  # 2 instrucciones por la comparación y asignación

        execints += 2  # Por las dos comparaciones en cada iteración

    execints += 1  # Incrementa por la línea de retorno final
    return biggest_1 * biggest_2  # Devuelve el producto máximo de los dos números más grandes




if __name__ == '__main__':
    n = int(input())  # Lee el tamaño de la lista
    input_line = input()  # Lee la línea de números
    numbers = list(map(int, input_line.split()))  # Convierte la línea de entrada en una lista de enteros
    max_pairwise_efficient(numbers)  # Llama a la función para calcular el producto máximo

    execints += 1  # Incrementa por el final de la ejecución de la función

    print("El numero de instrucciones es:", execints)  # Imprime el número de instrucciones ejecutadas



Writing max_pairwise_efficient.py


Calculate estimated time by instruction and number of instructions by second

In [84]:
!time python3 max_pairwise_efficient.py < input11.txt > pythonOutputSort11.txt
!python3 max_pairwise_efficient.py < input11.txt
num_inst_per_seg = 1200005/0.165
print(f"{num_inst_per_seg} tiempo por instrucción")


real	0m0.163s
user	0m0.141s
sys	0m0.020s
El numero de instrucciones es: 1200005
7272757.575757575 tiempo por instrucción


### Estimate and verify the maximun n that can be solved in 1 second

In [85]:
import time
import random
n = 0 #numero de instrucciones

def instruction(numbers):
    m = len(numbers)
    biggest_1 = float('-inf')
    biggest_2 = float('-inf')
    global n
    n += 3

    for i in range(m):
        new = numbers[i]
        temp = biggest_1
        n += 2
        if new > biggest_1:
            biggest_2 = biggest_1
            biggest_1 = new
            n += 2
        if temp > biggest_2:
            biggest_2 = temp
            n += 1
        n += 2
    return n
# Ejecutar la función y medir el tiempo de ejecución
start_time = time.time()
numbers = [random.randint(1, 100) for _ in range(10)]
instruction = instruction(numbers) # Suponiendo que 'numbers' es una lista válida
end_time = time.time()

# Calcular el número de instrucciones por segundo
elapsed_time = end_time - start_time
instructions_per_second = instruction / elapsed_time
#print(start_time)
#print(end_time)
print(f"Número de instrucciones por segundo: {instructions_per_second}")

Número de instrucciones por segundo: 358620.7881040892


## Efficient solution with largest and second-largest in an list, array or vector in C++ and Python

 $$T(n) < 10 n  = O(n)$$





Add in each line a coment with the number of intstrcution that are executed in that line

In [86]:
def efficientSolution(numbers):  # 1 instrucción para definir la función
    largest = 0  # 1 instrucción para asignar el valor 0 a largest
    secondLargest = 0  # 1 instrucción para asignar el valor 0 a secondLargest
    for i in numbers:  # 1 instrucción para iniciar el bucle
        if i > largest:  # 1 instrucción para la comparación i > largest
            secondLargest = largest  # 1 instrucción para asignar largest a secondLargest
            largest = i  # 1 instrucción para asignar i a largest
        elif i > secondLargest:  # 1 instrucción para la comparación i > secondLargest
            secondLargest = i  # 1 instrucción para asignar i a secondLargest

    result = secondLargest * largest  # 1 instrucción para la multiplicación y asignación a result
    return result  # 1 instrucción para retornar el resultado

n = int(input())  # 1 instrucción para leer la entrada y convertirla a entero
numbers = [int(i) for i in input().split()]  # 2 instrucciones: leer entrada y convertir a lista de enteros
print(efficientSolution(numbers))  # 1 instrucción para llamar a la función y mostrar el resultado


KeyboardInterrupt: Interrupted by user

Test with inputs 1 and 2



In [ ]:
!python3 efficientSolution.py < input1.txt
!python3 efficientSolution.py < input2.txt

Test and time it with input 3

In [ ]:
!time python3 efficientSolution.py < input3.txt

20 stress tests

In [87]:
import random
def efficientSolution(numbers):
    largest = 0
    secondLargest = 0
    for i in numbers:
        if i > largest:
            secondLargest = largest
            largest = i
        elif i > secondLargest:
            secondLargest = i

    return secondLargest * largest

def max_pairwise_product(numbers):
    max_product = 0
    n = len(numbers)

    for first in range(n):
        for second in range(n):
            if first != second:
                aux_product = numbers[first] * numbers[second]
                if aux_product > max_product:
                    max_product = aux_product

    return max_product

for _ in range(20):  # Run the test 10 times
    n = random.randint(1, 12)  # Use a large n
    numbers = [random.randint(1, 10**9) for _ in range(n)]  # Generate n random numbers between 1 and 10^9

    print("-------------------------------------------------")
    print(f"Testing with the list: {numbers}")
    print(f"The efficient solution gives the result {efficientSolution(numbers)}")
    print(f"The comparison solution gives the result {max_pairwise_product(numbers)}")
    if efficientSolution(numbers) == max_pairwise_product(numbers):
        print("The solution is correct OK")
    else:
        print("ERROR")
    print("-------------------------------------------------")




-------------------------------------------------
Testing with the list: [303049580, 20893774, 526024796, 798187025, 821806893, 730153795, 906567609, 588711377]
The efficient solution gives the result 745023510046728837
The comparison solution gives the result 745023510046728837
The solution is correct OK
-------------------------------------------------
-------------------------------------------------
Testing with the list: [185415732, 443064688, 144183992, 347919291]
The efficient solution gives the result 154150752116096208
The comparison solution gives the result 154150752116096208
The solution is correct OK
-------------------------------------------------
-------------------------------------------------
Testing with the list: [306532939, 864758814, 641917469]
The efficient solution gives the result 555103789178321766
The comparison solution gives the result 555103789178321766
The solution is correct OK
-------------------------------------------------
--------------------------

Count executed instructions in code and verify with inputs 1,2 and 3


In [88]:
%%writefile efficientSolutionexcec.py
exceintspy= 0
def efficientSolution(numbers):
    global exceintspy
    largest = 0
    exceintspy+=1
    secondLargest = 0
    exceintspy+=1
    exceintspy+=1
    for i in numbers:
        exceintspy+=2
        exceintspy+=1
        if i > largest:
            secondLargest = largest
            exceintspy+=1
            largest = i
            exceintspy+=1
        elif i > secondLargest:
            secondLargest = i
            exceintspy+=1
    exceintspy+=1
    result= secondLargest * largest
    exceintspy+=2 # multiplciation instruction
    return result

n = int(input())
exceintspy+=1
numbers = [int(i) for i in input().split()]
exceintspy+=n
print(efficientSolution(numbers))
exceintspy+=1
print(f"Execution Times: {exceintspy}")

Writing efficientSolutionexcec.py


Calculate estimated time by instruction and number of instructions by second

In [89]:
!time python3 efficientSolutionexcec.py< input11.txt > ptest.txt
!grep -xvFf output11.txt ptest.txt


real	0m0.165s
user	0m0.136s
sys	0m0.026s
Execution Times: 1200008


### Estimate and verify the maximun n that can be solved in 1 second

In [90]:
%%writefile efficientSolution.py
def efficientSolution(numbers):
    largest = 0                     #1
    secondLargest = 0               #1
    for i in numbers:               #2+2n
        if i > largest:
            secondLargest = largest
            largest = i
        elif i > secondLargest:
            secondLargest = i
    #WORST CASE 3n
    result= secondLargest * largest    #3
    return result

n = int(input())                            #1
numbers = [int(i) for i in input().split()] #n
print(efficientSolution(numbers))           #1

Writing efficientSolution.py


## Super efficient solution

Whitout storing the list in memory, only stpring the largest and second largest and whitout calling a function, add in each line a coment with the number of instructions that are executed in that line

in C++ and Python


 $$T(n) < 13 n  = O(n)$$


In [91]:
%%writefile max_pairwise_efficient.cpp

#include <iostream>
#include <limits>

long long max_pairwise_efficient() {
    int n;            //1
    std::cin >> n;    //1

    long long biggest_1 = std::numeric_limits<long long>::min(); //1
    long long biggest_2 = std::numeric_limits<long long>::min(); //1

    for (int i = 0; i < n; ++i) {  //2 + 2n
        int new_number;
        std::cin >> new_number;
        if (new_number > biggest_1) {
            biggest_2 = biggest_1;
            biggest_1 = new_number;
        } else if (new_number > biggest_2) {
            biggest_2 = new_number;
        }
    }

    return biggest_1 * biggest_2; //1
}

int main() {
    std::cout << max_pairwise_efficient() << std::endl; //1
    return 0; //1
}


Writing max_pairwise_efficient.cpp


In [92]:
!g++ max_pairwise_efficient.cpp -o max_pairwise_efficient -std=c++11
!./max_pairwise_efficient

^C


Test with inputs 1 and 2




In [93]:
!g++ max_pairwise_efficient.cpp -o max_pairwise_efficient -std=c++11
!./max_pairwise_efficient < input1.txt
!./max_pairwise_efficient < input2.txt

90
90


Test and time it with input 3

In [94]:
!g++ max_pairwise_efficient.cpp -o max_pairwise_efficient -std=c++11
!./max_pairwise_efficient < input3.txt

40000000000


20 stress tests

In [95]:
%%writefile stressTestmax_pairwise_efficient.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <random>
long long max_pairwise_efficient(const std::vector<int>& numbers) {
    int n = numbers.size(); //1

    long long biggest_1 = std::numeric_limits<long long>::min(); //1
    long long biggest_2 = std::numeric_limits<long long>::min(); //1

    for (int i = 0; i < n; ++i) { //2+2n
        int new_number = numbers[i];
        if (new_number > biggest_1) {
            biggest_2 = biggest_1;
            biggest_1 = new_number;
        } else if (new_number > biggest_2) { //2+2n
            biggest_2 = new_number;
        }
    }

    return biggest_1 * biggest_2;//1
}

int main() {
    // 20 stress tests
    for (int test = 1; test <= 20; ++test) { //2+2n
        int n = rand() % 11 + 2;  // Use a large n
        std::vector<int> numbers(n);

        // Generate n random numbers between 1 and 10^9
        for (int i = 0; i < n; ++i) {        //2+2n
            numbers[i] = rand() % 1000000000 + 1;
        }

        std::cout << "-------------------------------------------------" << std::endl; //1
        std::cout << "Testing with the list: "; //1
        for (int i = 0; i < n; ++i) {           //1
            std::cout << numbers[i] << " ";     //1
        }
        std::cout << std::endl;//1

        std::cout << "The efficient solution gives the result: " << max_pairwise_efficient(numbers) << std::endl; //1

        // Add any additional test or comparison here if needed

        std::cout << "-------------------------------------------------" << std::endl; //1
    }

    return 0;
}

Writing stressTestmax_pairwise_efficient.cpp


In [96]:
!g++ stressTestmax_pairwise_efficient.cpp -o stressTestmax_pairwise_efficient -std=c++11
!./stressTestmax_pairwise_efficient

-------------------------------------------------
Testing with the list: 846930887 681692778 714636916 957747794 424238336 719885387 649760493 596516650 
The efficient solution gives the result: 811146188694713278
-------------------------------------------------
-------------------------------------------------
Testing with the list: 25202363 350490028 783368691 
The efficient solution gives the result: 274562914442913348
-------------------------------------------------
-------------------------------------------------
Testing with the list: 44897764 967513927 365180541 540383427 304089173 
The efficient solution gives the result: 522828491542487829
-------------------------------------------------
-------------------------------------------------
Testing with the list: 35005212 521595369 
The efficient solution gives the result: 18258556470063228
-------------------------------------------------
-------------------------------------------------
Testing with the list: 726956430 33646

Count executed instructions in code and verify with inputs 1,2 and 3

In [97]:
%%writefile max_pairwise_efficient2.cpp
#include <iostream>
#include <vector>
#include <limits>

int execints = 0;

long long max_pairwise_efficient(const std::vector<int>& numbers) {
    int n = numbers.size(); //1
    long long biggest_1 = std::numeric_limits<long long>::min(); //1
    long long biggest_2 = std::numeric_limits<long long>::min(); //1
    execints += 3;

    for (int i = 0; i < n; ++i) { //2+2n
        int new_number = numbers[i];
        long long temp = biggest_1;
        if (new_number > biggest_1) {
            execints += 2;  // Contar la comparación y la actualización
            biggest_2 = biggest_1;
            biggest_1 = new_number;
        }
        if (temp > biggest_2) {
            execints += 2;  // Contar la comparación y la actualización
            biggest_2 = temp;
        }
        execints += 2;  // Contar las dos comparaciones
    }
    execints += 2;  // Contar el aumento final y el retorno

    return execints;

}

int main() {
    int n; //1
    std::cin >> n; //1
    std::vector<int> numbers(n); //1
    for (int i = 0; i < n; ++i) { //2+2n
        std::cin >> numbers[i];
    }
    long long result = max_pairwise_efficient(numbers); //1
    std::cout << "El numero de instrucciones es: " << execints << "\n"; //1
    return 0;
}

Writing max_pairwise_efficient2.cpp


In [98]:
%%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_efficient2.cpp -o max_pairwise_efficient2
for i in {1..3}
do
    echo "Input #$i"
    ./max_pairwise_efficient2< input$i.txt
done

Overwriting run_all_test.sh


In [99]:
!chmod u+x run_all_test.sh
!./run_all_test.sh


Input #1
El numero de instrucciones es: 45
Input #2
El numero de instrucciones es: 29
Input #3
El numero de instrucciones es: 25


Calculate estimated time by instruction and number of instructions by second

In [100]:
time_per_instruction = 0.41/1600007
time_per_instruction

2.5624887891115476e-07

### Super efficient solution in Python

Whitout storing the list in memory and whitout calling a function, add in each line a coment with the number of intstrcution that are executed in that line

In [101]:
%%writefile max_efficient.py

def max_pairwise_efficient():
    n = int(input())
    values = input().split()
    biggest_1 = float('-inf')
    biggest_2 = float('-inf')

    for new in map(int, values):
        temp = biggest_1
        if new > biggest_1:
            biggest_2 = biggest_1
            biggest_1 = new
        if temp > biggest_2:
            biggest_2 = temp

    return biggest_1 * biggest_2

if __name__ == '__main__':
    print(max_pairwise_efficient())

Writing max_efficient.py


Test with inputs 1 and 2




In [102]:
%%writefile max_efficient2.py

def max_pairwise_efficient():
    n = int(input())
    values = input().split()
    biggest_1 = float('-inf')
    biggest_2 = float('-inf')

    for new in map(int, values):
        temp = biggest_1
        if new > biggest_1:
            biggest_2 = biggest_1
            biggest_1 = new
        elif temp > biggest_2:
            biggest_2 = temp

    return biggest_1 * biggest_2

if __name__ == '__main__':
    print(max_pairwise_efficient())

Writing max_efficient2.py


In [103]:
%%writefile run_all_test2.sh
#comparing all the executions to check the differences between the outputs
for i in {1..2}
do
    python3 max_efficient2.py < input$i.txt > pythonOutput$i.txt
done

Writing run_all_test2.sh


In [104]:
!chmod u+x run_all_test2.sh
!./run_all_test.sh

Input #1
El numero de instrucciones es: 45
Input #2
El numero de instrucciones es: 29
Input #3
El numero de instrucciones es: 25


Test and time it with input 3

In [105]:
!time python3 max_efficient2.py < input3.txt > pythonOutput3.txt


real	0m0.038s
user	0m0.021s
sys	0m0.014s


20 stress tests

In [106]:
#%%writefile stressTestmax_pairwise_efficient.cpp
import random

def max_pairwise_efficient(numbers):
    n = len(numbers)
    biggest_1 = float('-inf')
    biggest_2 = float('-inf')

    for i in range(n):
        new = numbers[i]
        temp = biggest_1
        if new > biggest_1:
          biggest_2 = biggest_1
          biggest_1 = new
        if temp > biggest_2: biggest_2 = temp

    return biggest_1 * biggest_2

# 20 stress tests
for _ in range(20):
    n = random.randint(2, 12)  # Use a large n
    numbers = [random.randint(1, 10**9) for _ in range(n)]  # Generate n random numbers between 1 and 10^9

    print("-------------------------------------------------")
    print(f"Testing with the list: {numbers}")
    print(f"The efficient solution gives the result {max_pairwise_efficient(numbers)}")
    # Add any additional test or comparison here if needed
    print("-------------------------------------------------")

-------------------------------------------------
Testing with the list: [102938099, 916746048, 950520933, 418828843]
The efficient solution gives the result 903490044071190489
-------------------------------------------------
-------------------------------------------------
Testing with the list: [731626740, 297787584]
The efficient solution gives the result 535277686683027600
-------------------------------------------------
-------------------------------------------------
Testing with the list: [517935555, 21116598, 235098784, 787330595]
The efficient solution gives the result 407786508689805225
-------------------------------------------------
-------------------------------------------------
Testing with the list: [629700030, 926258817, 98567242, 14318658, 534998366]
The efficient solution gives the result 857955396070239489
-------------------------------------------------
-------------------------------------------------
Testing with the list: [302273264, 544339639, 266809538,

Count executed instructions in code and verify with inputs 1,2 and 3

In [107]:
%%writefile max_pairwise_efficient.py
import random
execints = 0

def max_pairwise_efficient(numbers):
    global execints
    n = len(numbers)
    biggest_1 = float('-inf')
    biggest_2 = float('-inf')
    execints += 3

    for i in range(n):
        new_number = numbers[i]
        temp = biggest_1
        execints += 2
        if new_number > biggest_1:
            biggest_2 = biggest_1
            biggest_1 = new_number
            execints += 2
        if temp > biggest_2:
            biggest_2 = temp
            execints += 2
        execints += 2

    execints += 1
    return biggest_1 * biggest_2

if __name__ == '__main__':
    n = int(input())
    input_line = input()
    numbers = list(map(int, input_line.split()))  # Divide la cadena en números y convierte cada uno a entero
    max_pairwise_efficient(numbers)
    execints += 1

    print("El numero de instrucciones es:", execints)

Overwriting max_pairwise_efficient.py


In [108]:
%%writefile run_all_test.sh
for i in {1..3}
do
    echo "Input #$i"
    python3 max_pairwise_efficient.py < input$i.txt
done

Overwriting run_all_test.sh


In [109]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

Input #1
El numero de instrucciones es: 65
Input #2
El numero de instrucciones es: 49
Input #3
El numero de instrucciones es: 37


Calculate estimated time by instruction and number of instructions by second

In [110]:
!time python3 max_pairwise_efficient.py < input11.txt > pythonOutputSort11.txt
!python3 max_pairwise_efficient.py < input11.txt
num_inst_per_seg = 1200005/0.165
print(f"{num_inst_per_seg} tiempo por instrucción")


real	0m0.162s
user	0m0.136s
sys	0m0.022s
El numero de instrucciones es: 1200005
7272757.575757575 tiempo por instrucción


### Estimate and verify the maximun n that can be solved in 1 second

In [111]:
import time
import random
n = 0 #numero de instrucciones

def instruction(numbers):
    m = len(numbers)
    biggest_1 = float('-inf')
    biggest_2 = float('-inf')
    global n
    n += 3

    for i in range(m):
        new = numbers[i]
        temp = biggest_1
        n += 2
        if new > biggest_1:
            biggest_2 = biggest_1
            biggest_1 = new
            n += 2
        if temp > biggest_2:
            biggest_2 = temp
            n += 1
        n += 2
    return n
# Ejecutar la función y medir el tiempo de ejecución
start_time = time.time()
numbers = [random.randint(1, 100) for _ in range(10)]
instruction = instruction(numbers) # Suponiendo que 'numbers' es una lista válida
end_time = time.time()

# Calcular el número de instrucciones por segundo
elapsed_time = end_time - start_time
instructions_per_second = instruction / elapsed_time
#print(start_time)
#print(end_time)
print(f"Número de instrucciones por segundo: {instructions_per_second}")

Número de instrucciones por segundo: 350743.1916376307


## Perform 100 randomized automated tests in C++ and Python with a porgram  compares the answers,for the same 100 random input instances, of the super efficient solution $(T(n) < 10 n = O(n))$ against the sorting solution  $(T(n)  = O(n \lg{n}))$,  if the answers are not the same stops and displays "Error" listing  the input and the two answers; if the 100 tests are completed without stoping displays "Pass".

#C++

In [112]:
%%writefile stressTest.cpp

#include <iostream>
#include <vector>
#include <algorithm>
#include <random>

long long max_pairwise_efficient(const std::vector<int>& numbers) { //stressTestC_efficien
    int n = numbers.size(); //1

    long long biggest_1 = std::numeric_limits<long long>::min(); //1
    long long biggest_2 = std::numeric_limits<long long>::min(); //1

    for (int i = 0; i < n; ++i) { //2+2n
        int new_number = numbers[i];
        if (new_number > biggest_1) {
            biggest_2 = biggest_1;
            biggest_1 = new_number;
        } else if (new_number > biggest_2) { //2+2n
            biggest_2 = new_number;
        }
    }

    return biggest_1 * biggest_2;//1
}

long long max_pairwise_product_sort(std::vector<int>& numbers) { //stressTestC_sort
    std::sort(numbers.rbegin(), numbers.rend());
    return static_cast<long long>(numbers[0]) * numbers[1];
}

int main() {
    std::random_device rd;
    std::mt19937 gen(rd());

    for (int i = 0; i < 100; ++i) {
        int n = std::uniform_int_distribution<int>(2, 12)(gen);  // Use a large n
        std::vector<int> numbers(n);
        for (int j = 0; j < n; ++j) {
            numbers[j] = std::uniform_int_distribution<int>(1, 1e9)(gen);  // Generate n random numbers between 1 and 10^9
        }

        std::cout << "-------------------------------------------------\n";
        std::cout << "Testing with the list: ";
        for (int num : numbers) {
            std::cout << num << " ";
        }
        std::cout << "\n";

        long long efficient_result = max_pairwise_efficient(numbers);
        long long comparison_result = max_pairwise_product_sort(numbers);  // Replace with your comparison function result

        std::cout << "The efficient solution gives the result " << efficient_result << "\n";
        std::cout << "The comparison solution gives the result " << comparison_result << "\n";

        if (efficient_result == comparison_result) {
            std::cout << "The solution is correct OK\n";
        } else {
            std::cout << "ERROR\n";
        }

        std::cout << "-------------------------------------------------\n";
    }

    return 0;
}

Writing stressTest.cpp


In [113]:
!g++ stressTest.cpp -o stressTest -std=c++11
!./stressTest

-------------------------------------------------
Testing with the list: 499953356 632662014 879941067 62624838 471183613 485269726 
The efficient solution gives the result 556705287649528938
The comparison solution gives the result 556705287649528938
The solution is correct OK
-------------------------------------------------
-------------------------------------------------
Testing with the list: 910139389 506693076 291780437 756986723 513138181 998162842 934378553 445625640 27548111 
The efficient solution gives the result 932661951966327626
The comparison solution gives the result 932661951966327626
The solution is correct OK
-------------------------------------------------
-------------------------------------------------
Testing with the list: 536522005 728689417 648430703 383838482 574403048 323891600 449285247 171621552 531433227 137169440 
The efficient solution gives the result 472504590933970151
The comparison solution gives the result 472504590933970151
The solution is cor

#Python

In [114]:
import random
def max_pairwise_product_sort(numbers):
  numbers.sort(reverse=True)
  return numbers[0] * numbers[1]

def max_pairwise_product(numbers):
    n = len(numbers)
    max_product = 0

    for first in range(n):
        for second in range(n):
            if first != second:
                aux_product = numbers[first] * numbers[second]

                if aux_product > max_product:
                  max_product = aux_product

    return max_product


for _ in range(20):  # Run the test 20 times
    n = random.randint(2, 12)  # Use a large n
    numbers = [random.randint(1, 10**9) for _ in range(n)]  # Generate n random numbers between 1 and 10^9

    print("-------------------------------------------------")
    print(f"Testing with the list: {numbers}")
    print(f"The efficient solution gives the result {max_pairwise_product_sort(numbers)}")
    print(f"The comparison solution gives the result {max_pairwise_product(numbers)}")
    if max_pairwise_product_sort(numbers) == max_pairwise_product(numbers):
        print("The solution is correct OK")
    else:
        print("ERROR")
    print("-------------------------------------------------")

-------------------------------------------------
Testing with the list: [754684391, 896219838, 160437189, 764313767, 42332974, 790540047, 558242653]
The efficient solution gives the result 708497672854852386
The comparison solution gives the result 708497672854852386
The solution is correct OK
-------------------------------------------------
-------------------------------------------------
Testing with the list: [407752971, 716880274, 248131913, 706850861, 463222471, 33254130, 538302400, 647368639, 464236858, 424847127]
The efficient solution gives the result 506727438910815914
The comparison solution gives the result 506727438910815914
The solution is correct OK
-------------------------------------------------
-------------------------------------------------
Testing with the list: [464211585, 442297809, 706808913, 40006347]
The efficient solution gives the result 328108885795857105
The comparison solution gives the result 328108885795857105
The solution is correct OK
------------